add `SetDutyCycle` mock

rursprung opened the pull request:
this is the 1.0 equivalent of the old `PwmPin` trait (the mock for this has been added in #52).

note that the test coverage is fully handled by the doc test which is why there's no additional `mod test` in this file.

this fixes #73

(note: this replaces #74 which was targeting the now defunct `1-alpha` branch)

rursprung commented:
it seems that the format of the assertion message is now different in nightly? i.e. the failure seems to be unrelated...
```
       panic message: `"assertion `left == right` failed: i2c::write unexpected mode\n  left: Read\n right: Write"`,
 expected substring: `"assertion failed: `(left == right)`\n  left: `Read`,\n right: `Write`: i2c::write unexpected mode"`
```

(however, i have the very odd experience that _without_ my PR it builds fine locally but with it it also fails with `cargo +nightly test --all-features`... i can't explain that right now 😕)

dbrgn commented:
Yeah, nightly apparently changed the assertion format. If we adapt to nightly, then tests on stable will fail.

I think I'll simply reduce the assertions to a smaller subset of the message.

dbrgn commented:
@rursprung should be fixed on `main`, could you rebase?

rursprung commented:
> @rursprung should be fixed on `main`, could you rebase?

rebased, works fine now. thanks for fixing this!

rursprung commented:
@dbrgn: while it wasn't strictly necessary (would've merged just fine also without this) i've now rebased this on top of the latest `main` and also adjusted a comment which still had a link to the e-h beta documentation.
once this PR lands in a release of e-h-m (it was my hope that it'd land in the release you just did) i'll be able to point to your release again in my crate (since i need this)

thanks for your work on this crate!

dbrgn commented on src/eh1/pwm.rs:
```diff
@@ -0,0 +1,130 @@
+//! use embedded_hal_mock::eh1::pwm::{Transaction as PwmTransaction, Mock as PwmMock};
+//! use eh1::pwm::SetDutyCycle;
+//!
+//! let err = MockError::Io(ErrorKind::NotConnected);
+//!
+//! // Configure expectations
+//! let expectations = [
+//!     PwmTransaction::get_max_duty_cycle(100),
+//!     PwmTransaction::set_duty_cycle(50),
+//!     PwmTransaction::set_duty_cycle(101).with_error(err.clone()),
+//! ];
```
Since the error is only used once, this would be less confusing to read:

```suggestion
//! // Configure expectations
//! let expectations = [
//!     PwmTransaction::get_max_duty_cycle(100),
//!     PwmTransaction::set_duty_cycle(50),
//!     PwmTransaction::set_duty_cycle(101).with_error(MockError::Io(ErrorKind::NotConnected)),
//! ];
```

dbrgn commented on src/eh1/pwm.rs:
```diff
@@ -0,0 +1,130 @@
+    SetDutyCycle(u16),
+}
+
+/// Mock SetDutyCycle implementation
```
```suggestion
/// Mock PWM `SetDutyCycle` implementation
```

dbrgn commented on src/eh1/pwm.rs:
```diff
@@ -0,0 +1,130 @@
+
+        assert_eq!(err, None, "error not supported by get_max_duty_cycle!");
+
+        if let TransactionKind::GetMaxDutyCycle(duty) = kind {
+            duty
+        } else {
+            unreachable!();
+        }
```
This branch is not actually unreachable, it could be `TransactionKind::SetDutyCycle(...)`. In that case, there _is_ a panic, but the assertion message is very user un-friendly ("unreachable code reached" instead of "expected x, found y".

I'd suggest this:

```suggestion
        match kind {
            TransactionKind::GetMaxDutyCycle(duty) => duty,
            other => panic!("expected get_max_duty_cycle, got {:?}", other),
        }
```

rursprung commented on src/eh1/pwm.rs:
```diff
@@ -0,0 +1,130 @@
+//! pwm.set_duty_cycle(101).expect_err("expected error return");
+//!
+//! pwm.done();
+//!
+//! // Update expectations
+//! pwm.expect(&[]);
+//! // ...
+//! pwm.done();
```
i guess that was copy & paste from `pin.rs` - just noticed that the others also don't have that.

rursprung commented:
thanks for all your feedback @dbrgn, i've incorporated it now

dbrgn reviewed (approved):
Thanks for the updates!

dbrgn commented:
Released in 0.10.0-rc.2.
